packages/html/src/events/mounted.rs

```diff
@@ -207,76 +207,219 @@ impl MountedData {
-    ///
-    /// ```rust, ignore
-    /// use dioxus::prelude::*;
-    /// use dioxus_web::WebEventExt; // provides [`as_web_event()`] method
-    ///
-    /// fn App() -> Element {
-    ///     rsx! {
-    ///         div {
-    ///             id: "some-id",
-    ///             onmounted: move |element| {
-    ///                 // You can use the web_event trait to downcast the element to a web specific event. For the mounted event, this will be a web_sys::Element
-    ///                 let web_sys_element = element.as_web_event();
-    ///                 assert_eq!(web_sys_element.id(), "some-id");
-    ///             }
-    ///         }
-    ///     }
-    /// }
-    /// ```
-    onmounted
-];
+// ============================================================================
+// Cleanup support for onmounted
+// ============================================================================
+
+/// A cleanup function to run when the element is unmounted.
+pub type MountedCleanup = Box<dyn FnOnce() + 'static>;
+
+thread_local! {
+    /// Storage for cleanup closures returned by onmounted handlers.
+    /// After firing a mounted event, the renderer should call `take_mounted_cleanup()`
+    /// to retrieve any cleanup closure that was registered.
+    static MOUNTED_CLEANUP: RefCell<Option<MountedCleanup>> = const { RefCell::new(None) };
+}
+
+/// Retrieve and clear any pending cleanup from the last mounted event.
+///
+/// Renderers should call this after firing a mounted event to capture
+/// any cleanup closure that was returned by the handler.
+pub fn take_mounted_cleanup() -> Option<MountedCleanup> {
+    MOUNTED_CLEANUP.with(|c| c.borrow_mut().take())
+}
+
+/// Trait to allow onmounted handlers to optionally return a cleanup closure.
+///
+/// This enables the pattern:
+/// ```rust,ignore
+/// onmounted: move |e| {
+///     start_animation(e.data());
+///     move || stop_animation(e.data())  // cleanup returned
+/// }
+/// ```
+///
+/// Handlers can return:
+/// - `()` - no cleanup
+/// - Any `FnOnce()` closure - cleanup to run on unmount
+/// - `async {}` block - spawned as task, no cleanup support
+pub trait SpawnIfAsyncWithCleanup<Marker>: Sized {
+    /// Process the return value, storing any cleanup closure and spawning async blocks.
+    fn spawn_with_cleanup(self);
+}
+
+// No cleanup - handler returns ()
+impl SpawnIfAsyncWithCleanup<()> for () {
+    fn spawn_with_cleanup(self) {
+        // No cleanup needed
+    }
+}
+
+/// Marker for cleanup closures
+#[doc(hidden)]
+pub struct CleanupMarker;
+
+// Handler returns a cleanup closure
+impl<F: FnOnce() + 'static> SpawnIfAsyncWithCleanup<CleanupMarker> for F {
+    fn spawn_with_cleanup(self) {
+        MOUNTED_CLEANUP.with(|c| *c.borrow_mut() = Some(Box::new(self)));
```

Comment:
Instead of a thread local, can you add the cleanup callback to the mounted event? This variant would need to do something like `event.data().set_on_cleanup(self)`